src/coreclr/jit/morph.cpp

```diff
@@ -12500,6 +12500,20 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
 #endif // !TARGET_64BIT
             break;
 
+        case GT_ARR_LENGTH:
+            if (op1->OperIs(GT_CNS_STR))
+            {
+                // Optimize `ldstr + String::get_Length()` to CNS_INT
+                // e.g. "Hello".Length => 5
+                GenTreeIntCon* iconNode = impStringLiteralLength(op1->AsStrCon());
```

Comment:
Seems odd to call a function from morph that is "impX", although I guess it wouldn't be the first case.

Reply:
Yeah there are already quite a few:
```
impGetRefAnyClass
impInlineRoot
impTailCallRetTypeCompatible
impFixupStructReturnType
impCheckForNullPointer
impIsAddressInLocal
impReadyToRunHelperToTree
```

Reply:
Even so, it seems like it belongs over in gentree.cpp.